pkg/sentry/socket/socket.go

```diff
@@ -375,3 +376,191 @@ func UnmarshalSockAddr(family int, data []byte) linux.SockAddr {
 		panic(fmt.Sprintf("Unsupported socket family %v", family))
 	}
 }
+
+// ifconfIoctl populates a struct ifconf for the SIOCGIFCONF ioctl.
+func ifconfIoctl(ctx context.Context, io usermem.IO, ifc *linux.IFConf) error {
+	stack := inet.StackFromContext(ctx)
+	if stack == nil {
+		return syserr.ErrNoDevice.ToError()
+	}
+
+	// If Ptr is NULL, return the necessary buffer size via Len.
+	// Otherwise, write up to Len bytes starting at Ptr containing ifreq
+	// structs.
+	if ifc.Ptr == 0 {
+		ifc.Len = int32(len(stack.Interfaces())) * int32(linux.SizeOfIFReq)
+		return nil
+	}
+
+	max := ifc.Len
+	ifc.Len = 0
+	for key, ifaceAddrs := range stack.InterfaceAddrs() {
+		iface := stack.Interfaces()[key]
+		for _, ifaceAddr := range ifaceAddrs {
+			// Don't write past the end of the buffer.
+			if ifc.Len+int32(linux.SizeOfIFReq) > max {
+				break
+			}
+			if ifaceAddr.Family != linux.AF_INET {
+				continue
+			}
+
+			// Populate ifr.ifr_addr.
+			ifr := linux.IFReq{}
+			ifr.SetName(iface.Name)
+			usermem.ByteOrder.PutUint16(ifr.Data[0:2], uint16(ifaceAddr.Family))
+			usermem.ByteOrder.PutUint16(ifr.Data[2:4], 0)
+			copy(ifr.Data[4:8], ifaceAddr.Addr[:4])
+
+			// Copy the ifr to userspace.
+			dst := uintptr(ifc.Ptr) + uintptr(ifc.Len)
+			ifc.Len += int32(linux.SizeOfIFReq)
+			if _, err := usermem.CopyObjectOut(ctx, io, usermem.Addr(dst), ifr, usermem.IOOpts{
+				AddressSpaceActive: true,
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// IfconfIoctl returns a list of interface addresses or the buffer size
+// necessary to hold the list.
+func IfconfIoctl(ctx context.Context, io usermem.IO, addr usermem.Addr) error {
+	var ifc linux.IFConf
+
+	if _, err := usermem.CopyObjectIn(ctx, io, addr, &ifc, usermem.IOOpts{
+		AddressSpaceActive: true,
+	}); err != nil {
+		return err
+	}
+
+	if err := ifconfIoctl(ctx, io, &ifc); err != nil {
+		return err
+	}
+
+	_, err := usermem.CopyObjectOut(ctx, io, addr, ifc, usermem.IOOpts{
+		AddressSpaceActive: true,
+	})
+
+	return err
+}
+
+// InterfaceIoctl implements interface requests; and it returns corresponding interface
+// for customzation.
+func InterfaceIoctl(ctx context.Context, io usermem.IO, arg int, ifr *linux.IFReq) (inet.Interface, *syserr.Error) {
```

Comment:
I am worried that a shared netstack and hostinet implementation will make it harder to add these features to netstack given that we can't add them to hostinet.

Reply:
Let's come back to this after we have a conclusion on the above discussion on security.